cmd/kubeadm/app/util/marshal.go

```diff
@@ -52,6 +54,25 @@ func MarshalToYamlForCodecs(obj runtime.Object, gv schema.GroupVersion, codecs s
 	return runtime.Encode(encoder, obj)
 }
 
+// MarshalTo marshals data to the specified output format (JSON or YAML)
+func MarshalTo(data interface{}, outputFormat string) (string, error) {
+	of := strings.ToLower(outputFormat)
+	if of != "text" && of != "json" && of != "yaml" {
+		return "", errors.Errorf("invalid output format: %s", outputFormat)
+	}
+	bytes, err := json.MarshalIndent(data, "", "  ")
+	if err != nil {
+		return "", err
+	}
+	if outputFormat == "yaml" {
+		bytes, err = yaml.JSONToYAML(bytes)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to convert JSON output to YAML")
+		}
+	}
+	return string(bytes), nil
```

Comment:
i think we want the function to also handle "text", in which case just pass the intput -> output transparently?

Reply:
Thank you for the review.
not sure about it, but it might make sense. will try.

Reply:
> i think we want the function to also handle "text", in which case just pass the intput -> output transparently?

That would be hard to do as we pass data structure for marshalling and its text representation depends on the structure.

Reply:
may be I misunderstood what you mean here. I thought you want to handle converting structure to text form in MarshalToOutputFormat. Please, elaborate.

Reply:
@bart0sh normally before sending PRs, the way we do things in kubeadm is to create a proposal in a Google doc to explore the idea and find potential problems...

- we need to define a unified way that handles all possible output format cases in a utility function.
- we don't want to handle `if format == "text"` in a lot of places.
- we want to define all outputFormats as constants.
```  
OutputFormatText = "text"
OutputFormatJSON = "json"
OutputFormatYAML = "yaml"
```
- possibly the location of this should be `cmd/kubeadm/app/util/machinereadable.go`
- we can do marshaling and printing in the same function:
```
PrintMachineReadable(target, input *bytes.Buffer, data interface{}, outputFormat string) error {
   if outputFormat == "text" {
      if input == nil {
         return errors.Errorf("PrintMachineReadable input is nil for outputFormat: %s", outputFormat)
      }
       fmt.Fprintf(target, input.String())
       return nil
   }
   // marshall cases here in a utility function that can be unit tested.
}

// usage:
var buf bytes.Buffer
// write images to buf for format "text"
PrintMachineReadable(os.Stdout, &buf, someStruct, format)
```

it is still unclear to me how to handle commands like `init`, `join`, `reset` and `upgrade`.

Reply:
@neolit123 Thank you for the explanations. It looks like I underestimated complexity of the task. Do you think I still should create a proposal first? Can you point me out to other proposals that I can use as an example?

Reply:
i don't think it's that complex, possibly we just need to plan a little either in the machine readable ticket https://github.com/kubernetes/kubeadm/issues/494

i think a public google doc with the ideas will be a good start.